pkg/controller/atlasproject/custom_roles.go

```diff
@@ -0,0 +1,207 @@
+
+	for _, currentCustomRole := range currentCustomRoles {
+		if _, ok := desiredCustomRolesByName[currentCustomRole.Name]; !ok {
+			toDelete = append(toDelete, currentCustomRole)
+		}
+	}
+
+	ctx.Log.Debugw("Custom Roles to be deleted", "NumItems", len(toDelete))
+
+	var err error
+	for _, customRole := range toDelete {
+		_, delErr := ctx.Client.CustomDBRoles.Delete(context.Background(), projectID, customRole.Name)
+
+		if delErr != nil {
+			err = multierror.Append(err, fmt.Errorf("failed to delete custom role \"%s\": %w", customRole.Name, delErr))
+		}
+
+		ctx.Log.Debugw("Removed Custom Role from current AtlasProject", "custom role", customRole.Name)
+	}
+
+	return err
+}
+
+func updateCustomRoles(ctx *workflow.Context, projectID string, currentCustomRoles []v1.CustomRole, desiredCustomRoles []v1.CustomRole) error {
+	toUpdate := make([]v1.CustomRole, 0, len(desiredCustomRoles))
+	currentCustomRolesByName := mapCustomRolesByName(currentCustomRoles)
+
+	for _, desiredCustomRole := range desiredCustomRoles {
+		customRole, ok := currentCustomRolesByName[desiredCustomRole.Name]
+
+		if !ok {
+			continue
+		}
+
+		if d := cmp.Diff(desiredCustomRole, customRole, cmpopts.EquateEmpty()); d != "" {
+			toUpdate = append(toUpdate, desiredCustomRole)
+		}
+	}
+
+	ctx.Log.Debugw("Custom Roles to be updated", "NumItems", len(toUpdate))
+
+	var err error
+	for _, customRole := range toUpdate {
+		data := customRole.ToAtlas()
+		// Patch fails when sending the role name in the body, needs clarification with cloud team
+		data.RoleName = ""
+		_, _, upErr := ctx.Client.CustomDBRoles.Update(context.Background(), projectID, customRole.Name, data)
+
+		if upErr != nil {
+			err = multierror.Append(err, fmt.Errorf("failed to update custom role \"%s\": %w", customRole.Name, upErr))
+		}
+
+		ctx.Log.Debugw("Updated Custom Role in current AtlasProject", "custom role", customRole.Name)
+	}
+
+	return err
+}
+
+func createCustomRoles(ctx *workflow.Context, projectID string, currentCustomRoles []v1.CustomRole, desiredCustomRoles []v1.CustomRole) error {
+	toCreate := make([]v1.CustomRole, 0, len(desiredCustomRoles))
+	currentCustomRolesByName := mapCustomRolesByName(currentCustomRoles)
+
+	for _, desiredCustomRole := range desiredCustomRoles {
+		if _, ok := currentCustomRolesByName[desiredCustomRole.Name]; !ok {
+			toCreate = append(toCreate, desiredCustomRole)
+		}
+	}
+
+	ctx.Log.Debugw("Custom Roles to be added", "NumItems", len(toCreate))
+
+	var err error
+	for _, customRole := range toCreate {
+		_, _, creErr := ctx.Client.CustomDBRoles.Create(context.Background(), projectID, customRole.ToAtlas())
+
+		if creErr != nil {
+			err = multierror.Append(err, fmt.Errorf("failed to create custom role \"%s\": %w", customRole.Name, creErr))
+		}
+
+		ctx.Log.Debugw("Created Custom Role in current AtlasProject", "custom role", customRole.Name)
+	}
+
+	return err
+}
+
+func mapCustomRolesByName(customRoles []v1.CustomRole) map[string]v1.CustomRole {
```

Comment:
I'm not sure that using the name as the identifier is a good idea. For example, if I create a project with such spec:
```
spec:
  customRoles:
    - actions:
        - name: LIST_SESSIONS
          resources:
            - cluster: true
        - name: KILL_ANY_SESSION
          resources:
            - cluster: true
      inheritedRoles:
        - database: admin
          name: enableSharding
        - database: admin
          name: backup
      name: ShardingAdmin
    - inheritedRoles:
        - database: test
          name: readWrite
        - database: test
          name: dbAdmin
      name: ShardingAdmin
  maintenanceWindow: {}
  name: Test project 1
  withDefaultAlertsSettings: true
```
The operator will create only one role.

Reply:
@Sugar-pack @helderjs  I think it make sense to mention that the role names should be unique. In this case the operator may show an error in the status, saying duplicates are not allowed.

Reply:
@Sugar-pack @igor-karpukhin As the name is unique and it's the identifier, wouldn't be better to add this rule as validation?